From b8a305b2a7af3d5151463f8edaa36aca0875e8de Mon Sep 17 00:00:00 2001
From: wh <wanghang@rock-chips.com>
Date: Sun, 15 Nov 2020 16:12:50 +0800
Subject: [PATCH 23/26] MediaProvider: modify to support soundrecorder app

1. soundrecorder can add record file to playlist
2. fix the record file type changed after reboot

Signed-off-by: wh <wanghang@rock-chips.com>
Change-Id: I9b5b6d3f3d3802437557ab88277eaa70a74507d6
---
 src/com/android/providers/media/MediaProvider.java       | 6 +++++-
 .../android/providers/media/scan/ModernMediaScanner.java | 9 ++++++++-
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 57a8d6ef..f69f3f2d 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -3287,6 +3287,9 @@ public class MediaProvider extends ContentProvider {
                 if (isCallingPackageSelf() || isCallingPackageLegacyWrite()) {
                     // Mutation allowed
                 } else {
+                    if ("com.android.soundrecorder".equals(getCallingPackageOrSelf())) {
+                        continue;
+                    }
                     Log.w(TAG, "Ignoring mutation of  " + column + " from "
                             + getCallingPackageOrSelf());
                     initialValues.remove(column);
@@ -7663,7 +7666,8 @@ public class MediaProvider extends ContentProvider {
     @Deprecated
     @VisibleForTesting
     public int getCallingPackageTargetSdkVersion() {
-        if ("com.android.music".equals(getCallingPackageOrSelf())) {
+        if ("com.android.music".equals(getCallingPackageOrSelf())
+            || "com.android.soundrecorder".equals(getCallingPackageOrSelf())) {
             return 26;
         }
         return mCallingIdentity.get().getTargetSdkVersion();
diff --git a/src/com/android/providers/media/scan/ModernMediaScanner.java b/src/com/android/providers/media/scan/ModernMediaScanner.java
index 58e63cd1..09513db3 100644
--- a/src/com/android/providers/media/scan/ModernMediaScanner.java
+++ b/src/com/android/providers/media/scan/ModernMediaScanner.java
@@ -585,7 +585,7 @@ public class ModernMediaScanner implements MediaScanner {
             queryArgs.putInt(MediaStore.QUERY_ARG_MATCH_FAVORITE, MediaStore.MATCH_INCLUDE);
             final String[] projection = new String[] {FileColumns._ID, FileColumns.DATE_MODIFIED,
                     FileColumns.SIZE, FileColumns.MIME_TYPE, FileColumns.MEDIA_TYPE,
-                    FileColumns.IS_PENDING};
+                    FileColumns.IS_PENDING, MediaStore.MediaColumns.OWNER_PACKAGE_NAME};
 
             final Matcher matcher = FileUtils.PATTERN_EXPIRES_FILE.matcher(realFile.getName());
             // If IS_PENDING is set by FUSE, we should scan the file and update IS_PENDING to zero.
@@ -600,6 +600,13 @@ public class ModernMediaScanner implements MediaScanner {
                     final String mimeType = c.getString(3);
                     final int mediaType = c.getInt(4);
                     isPendingFromFuse &= c.getInt(5) != 0;
+                    //make soundrecorder not change media_type
+                    final String owner_package_name = c.getString(6);
+                    if ("com.android.soundrecorder".equals(owner_package_name)
+                        && null != mimeType) {
+                        actualMimeType = mimeType;
+                        actualMediaType = mediaType;
+                    }
 
                     // Remember visiting this existing item, even if we skipped
                     // due to it being unchanged; this is needed so we don't
-- 
2.35.1

